app/gui/view/graph-editor/src/builtin/visualization/java_script/table.js

```diff
@@ -95,341 +49,197 @@ class TableVisualization extends Visualization {
+            rowData = parsedData.json.map((row, i) => ({ ['#']: i, value: toRender(row) }))
+            dataTruncated = parsedData.all_rows_count !== parsedData.json.length
         } else if (parsedData.json !== undefined) {
-            const table = genTable(parsedData.json, 0, undefined)
-            tabElem.innerHTML = style + table
+            columnDefs = [{ field: 'value' }]
+            rowData = [{ value: toRender(parsedData.json) }]
         } else {
-            const table = genDataframe(parsedData)
-            let suffix = ''
-            const allRowsCount = parsedData.all_rows_count
-            if (allRowsCount !== undefined) {
-                const includedRowsCount = parsedData.data.length > 0 ? parsedData.data[0].length : 0
-                const hiddenCount = allRowsCount - includedRowsCount
-                if (hiddenCount > 0) {
-                    let rows = 'rows'
-                    if (hiddenCount === 1) {
-                        rows = 'row'
-                    }
-                    suffix =
-                        '<span class="hiddenrows">&#8230; and ' +
-                        hiddenCount +
-                        ' more ' +
-                        rows +
-                        '.</span>'
+            const indices_header = (parsedData.indices_header ? parsedData.indices_header : []).map(
+                h => {
+                    const headerName = h === '#' ? 'Row#' : h
+                    return { field: h, headerName: headerName }
+                }
+            )
+            columnDefs = [...indices_header, ...parsedData.header.map(h => ({ field: h }))]
+
+            const rows =
+                parsedData.data && parsedData.data.length > 0
+                    ? parsedData.data[0].length
+                    : parsedData.indices && parsedData.indices.length > 0
+                    ? parsedData.indices[0].length
+                    : 0
+            rowData = Array.apply(null, Array(rows)).map((_, i) => {
+                const row = {}
+                const shift = parsedData.indices ? parsedData.indices.length : 0
+                columnDefs.map(
+                    (h, j) =>
+                        (row[h.field] = toRender(
+                            j < shift ? parsedData.indices[j][i] : parsedData.data[j - shift][i]
+                        ))
+                )
+                return row
+            })
+
+            dataTruncated = parsedData.all_rows_count !== rowData.length
+        }
+
+        // If the table contains more rows than an upper limit, the engine will send only some of all rows.
+        // If data is truncated, we cannot rely on sorting/filtering so will disable.
+        // A pinned row is added to tell the user the row count and that filter/sort is disabled.
+        if (dataTruncated) {
+            columnDefs[0].colSpan = p => {
+                if (p.data['__COL_SPAN__'] !== undefined) {
```

Comment:
the `'__COL_SPAN__'` literal appears multiple times in the code. I think its worth extracting it to a well-named constant.